[clang][Sema] Make format size estimator aware of %p's existence in format string

hazohelet opened the pull request:
GCC stops counting format string's size when it sees %p format in order to avoid `Wformat-truncation` false positive against Linux kernel's format extension (%pOF, for example).
This change makes clang's behavior align with GCC's.
As requested in https://github.com/llvm/llvm-project/issues/64871

nickdesaulniers commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -860,6 +860,16 @@ class EstimateSizeFormatHandler
 
+  bool ShouldStopOnFormatSpecifier(const analyze_printf::PrintfSpecifier &FS,
+                                   unsigned RemainLen) override {
+    if (FS.getConversionSpecifier().getKind() ==
+        analyze_printf::PrintfConversionSpecifier::pArg) {
+      assert(Size >= RemainLen && "no underflow");
+      Size -= RemainLen;
+      return true;
+    }
+    return false;
```
Might be shorter to invert the condition and return early.

nickdesaulniers commented on clang/lib/AST/PrintfFormatString.cpp:
```diff
@@ -429,7 +429,13 @@ bool clang::analyze_format_string::ParsePrintfString(FormatStringHandler &H,
       continue;
-    // We have a format specifier.  Pass it to the callback.
+    // We have a format specifier.
+    // In case the handler needs to abort parsing upon specific specifiers
```
End comment with correct punctuation.

nickdesaulniers reviewed (commented):
Thanks for the patch!

Can we document this behavior somewhere?

zygoloid commented:
Is there some way we can narrow the scope of this patch so we don't lose warnings for normal `snprintf`s, only for the kernel one? If we really can't tell the difference from the source code, we could move the affected warnings to a different warning group instead, so the kernel can turn them off.

hazohelet commented:
> Is there some way we can narrow the scope of this patch so we don't lose warnings for normal `snprintf`s, only for the kernel one? If we really can't tell the difference from the source code, we could move the affected warnings to a different warning group instead, so the kernel can turn them off.

Thanks for the feedback. I narrowed the scope by checking whether the next character of `%p` can be part of the kernel extended format specifier.
If people write `%pM` or a similar outside kernel, it still aborts the format size estimator and could lead to false negative. But I think it would hurt people much less because people usually put non-alphanumeric character after `%p`.

nickdesaulniers commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -163,6 +163,10 @@ Improvements to Clang's diagnostics
   result in string truncation.
   (`#64871: <https://github.com/llvm/llvm-project/issues/64871>`_).
+  Clang stops calculating the minimum length of the output format string upon %p
+  if the succeeding character can be a part of the format specifier in linux's
+  format extension in order to avoid false positive warnings against the linux
+  code base.
```
perhaps worth referring to "the Linux kernel's ..." rather than just "linux".

nickdesaulniers commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -851,6 +851,50 @@ class ScanfDiagnosticFormatHandler
+                                                       const char *E) {
+  assert(I < E && "format string not yet exhausted");
+  // Kernel Document: https://docs.kernel.org/core-api/printk-formats.html
+  switch (*I) {
```
The URL with anchor https://docs.kernel.org/core-api/printk-formats.html#pointer-types may be more precise to use in the comment.

nickdesaulniers replied:
I think it would also be helpful to say something along the lines of:

see the `pointer()` function in https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/lib/vsprintf.c.

nickdesaulniers commented:
> Is there some way we can narrow the scope of this patch so we don't lose warnings for normal `snprintf`s, only for the kernel one? If we really can't tell the difference from the source code, we could move the affected warnings to a different warning group instead, so the kernel can turn them off.

@zygoloid is this approach acceptable?

One thing I do worry about not having an explicit flag; we support clang-11+ for building the Linux kernel.

Let's say the kernel adds a new extension format specifiers `%pn`.  We need to fix up clang to recognize that now, too.  Then we fix clang in say clang-19.  clang-18 still warns and we can't turn it off.

Perhaps it would be better to have the flag to disable `%p` without disabling the rest of `-Wfortify-source`, than try to keep these two in sync.

zygoloid commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -851,6 +851,50 @@ class ScanfDiagnosticFormatHandler
+  }
+}
+
 class EstimateSizeFormatHandler
```
Here's what I'd suggest:

- Add a flag to this class to track whether we've seen any specifiers that the Linux kernel gives unusual behavior to.
- Set the flag after line 933 (`   case analyze_format_string::ConversionSpecifier::pArg:`).
- On line 1251 (`DiagID = diag::warn_fortify_source_format_overflow;`), set `DiagID` to a different value that's in a separate diagnostic group with a different `-W` flag that's nested within the `FortifySource` group.

That'll mean that no-one loses any diagnostic quality, and we have a new flag, say `-Wno-fortify-source-non-kprintf` that the Linux kernel can use to turn off these warnings.

zygoloid reviewed (dismissed):
This looks like the wrong approach to me. I really think you should be splitting the warning into two parts, so that only the kernel developers see a change in diagnostic output. That would seem to make this patch simpler as well as avoiding a (probably unimportant) regression for all other Clang users.

hazohelet commented:
Thanks for the detailed suggestion! I agree it's the appropriate approach.

I was planning to separate `Wformat-truncation` from `Wfortify-source` sometime after this PR (as was requested in the same issue), but now that we are going to introduce a new warning group, we should finish the separation before proceeding further on this.
Sadly we don't have any way to make stacked patches as of now, so I'll merge the separation change to this PR. If reviewers think it's unreasonable, let's make another PR for the separation then.

hazohelet committed:
[clang][Diagnostics] Separate Wformat-overflow and Wformat-truncation from Wfortify-source

Newly introduces `Wformat-overflow` and `Wformat-truncation` and imports the existing warning from `Wfortify-source` that correspond to GCC's counterpart so that they can be disabled separately from `Wfortify-source`.

hazohelet committed:
Introduce `-Wformat-overflow-non-kprintf` and `-Wformat-truncation-non-kprintf`  for format string that contains `%p`

zygoloid commented on clang/include/clang/Basic/DiagnosticGroups.td:
```diff
@@ -961,10 +961,16 @@ def FormatNonStandard : DiagGroup<"format-non-iso">;
 def FormatPedantic : DiagGroup<"format-pedantic">;
 def FormatTypeConfusion : DiagGroup<"format-type-confusion">;
+
+def FormatOverflowNonKprintf: DiagGroup<"format-overflow-non-kprintf">;
+def FormatOverflow: DiagGroup<"format-overflow", [FormatOverflowNonKprintf]>;
+def FormatTruncationNonKprintf: DiagGroup<"format-truncation-non-kprintf">;
+def FormatTruncation: DiagGroup<"format-truncation", [FormatTruncationNonKprintf]>;
```
Should these be added to the `FortifySource` group in addition to adding them to `Format`?

hazohelet replied:
Thanks. It makes sense to add this to `FortifySource` for backward compatibility.

zygoloid commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -1350,10 +1360,17 @@ void Sema::checkFortifiedBuiltinMemoryFunction(FunctionDecl *FD,
             llvm::APSInt::getUnsigned(H.getSizeLowerBound())
                 .extOrTrunc(SizeTypeWidth);
         if (FormatSize > *SourceSize && *SourceSize != 0) {
-          DiagID = diag::warn_fortify_source_format_truncation;
-          DestinationSize = SourceSize;
-          SourceSize = FormatSize;
-          break;
```
This diagnostic used to override the production of the `warn_fortify_source_size_mismatch` warning, and now is produced in addition to it. Is that intentional?

hazohelet replied:
This is intentional.
When the truncation warning was grouped under `Wfortify-source`, I thought it made sense to override the warning because there was no way to disable only one of `warn_fortify_source_size_mismatch` and the truncation warning.
Now that `Wformat-truncation` is made, I no longer think this warning should hide the `Wfortify-source` warning.

zygoloid reviewed (commented):
I'm happy with the overall approach here; please don't block on further review from me.

nickdesaulniers commented:
@nathanchance and I have done a bunch of testing of this revision (see comments https://github.com/ClangBuiltLinux/linux/issues/1923#issuecomment-1718144462 and down); this looks like something that will work nicely for the Linux kernel's use case! Thanks for all of the work here @hazohelet !

hazohelet committed:
Make `-Wformat-overflow` and `-Wformat-truncation` grouped in
`-Wfortify-source` as well as `-Wformat`

hazohelet committed:
[NFC] fix typo in release note

nickdesaulniers reviewed (approved):
Thanks for the fix!

zygoloid reviewed (approved):
Thank you!

nickdesaulniers commented:
ping @hazohelet . This should be good to land and should help clear some red in our CI for Linux kernel builds.

hazohelet commented:
> ping @hazohelet . This should be good to land and should help clear some red in our CI for Linux kernel builds.

Thanks for the ping. I was away for a while. I'm merging this after updating the PR descriptions.
